### ISSUE TITLE
[HOLD for payment 2023-11-29] [$1000] Removed reactions on offline mode re appear on the receiver account and there is no way to remove them

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#23679]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Open two new accounts on ND
2. Start a chat with each other
3. Go offline (disconnecting from the device required)
4. React with :+1: on account A to the last message posted
5. React with :+1: on account B to the last  message
6. Remove both the reactions from account A and account B
7. Go online
## Expected Result: 
There should be no reaction on the last message
## Actual Result: 
Account B shows the :+1: reaction on the last message sent, it says that account A reacted on it, but when opening account A the reaction is not there.

## Workaround:
Can the user still use Expensify without this being fixed? Have you informed them of the workaround?

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.46-1
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/43996225/b3a57aed-c325-48c1-916d-1c5ea9131190

https://github.com/Expensify/App/assets/43996225/66d4f42e-2093-43f4-917c-d9880e47b108

**Expensify/Expensify Issue URL:**
**Issue reported by:** @nathan-mulugeta
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1690268434954109

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~015cd66a26c22a758e</li>
        <li>Upwork Job ID: 1694287577039745024</li>
        <li>Last Price Increase: 2023-11-08</li>
<li>Automatic offers: </li>
<ul>
<li>aimane-chnaif | Reviewer | 27582689</li>
<li>barros001 | Contributor | 27582692</li>
<li>nathan-mulugeta | Reporter | 27582695</li>
</ul></li>
    </ul>
</details>

**Comment by StevenKKC** [comment]
Dupe of https://github.com/Expensify/App/issues/23388

**Comment by mallenexpensify** [comment]
Was able to reproduce, I fully turned off my wifi when doing so.   For the user that is still showing as having sent the 👍, they can click the 👍 emoji twice to remove it (if that matters).

**Comment by Nathan-Mulugeta** [comment]
Will do as soon as I can.

**Comment by Nathan-Mulugeta** [comment]
Hey @aimane-chnaif I just retested and I still am able to reproduce.

**Comment by mallenexpensify** [comment]
This bug never got to see external 👀 because it was potentially a dupe of https://github.com/Expensify/App/issues/23388. That one has been closed for a while (as non reproducible) and this one is reproducible so lets see if we can get some proposals to fix it.

**Comment by ShogunFire** [comment]
I didn't managed to pinpoint it because I don't see the pusher events but it feels like a backend problem. Maybe the pusher sends the message to add an emoji to the wrong user for some reason. Also something we can consider is using set instead of merge for the reactions in openReport, right now we have a merge on the whole action and I think merge on collections don't remove from the collection

**Comment by mallenexpensify** [comment]
@cead22 , added `Engineering` to see if you have knowledge about how pusher might be affecting this bug
> Maybe the pusher sends the message to add an emoji to the wrong user for some reason. 

@stephanieelliott  I'm off this week, can you please keep 👀 on this then I'll snag it back on Monday?  Thx

**Comment by cead22** [comment]
> Maybe the pusher sends the message to add an emoji to the wrong user for some reason.

That seems unlikely to me, and if it's happening, it's probably due to a bug on our end when calling Pusher. But I don't think that's the first thing to look at.

I think the first thing we should look at is:
1. The network queue when the reactions are added, and removed
2. The network requests when coming back online, and the order they're sent
3. The network responses

@mallenexpensify can you reproduce again and look into the above?
1. You can check using dev tools > application > storage > indexedDB > OnyxDB > keyvaluepairs > networkRequestQueue
2. You can check using dev tools > network. You can click on the request on the left and check payload for the request payload, and preview for the responses

<img width="1751" alt="image" src="https://github.com/Expensify/App/assets/165133/e0332ac4-10e5-4a78-ad4b-ef05d5bad13f">
<img width="1751" alt="image" src="https://github.com/Expensify/App/assets/165133/af378ff4-5fb5-497a-9781-8fdb67ce4243">

Also, can you check if this can be reproduced using different devices, or if it's only when using the same device (eg, desktop web in the issue description) while being logged in into two different accounts?

**Comment by ShogunFire** [comment]
Yes I was thinking of this also I had asked this on slack: https://expensify.slack.com/archives/C01GTK53T8Q/p1693669536590719?thread_ts=1693669536.590719&cid=C01GTK53T8Q I don't really understand what guarantees that the messages made by the pusher will be received in the same order they have been sent

**Comment by mallenexpensify** [comment]
@cead22 , took me a min (or... many),  
For user B, which has the 👍 'stuck' I'm seeing this
<img width="887" alt="image" src="https://github.com/Expensify/App/assets/22508304/73b95600-cbab-432a-99f9-5e0f9d587ee3">

Seeing this for network requests
<img width="872" alt="image" src="https://github.com/Expensify/App/assets/22508304/8bb0757e-1d4d-42f7-8b23-7c620fce9230">


Let me know if that doesn't look _right_, I went down a bit of an 🐰🕳 and the internet was going in/out while I was also going on/offline trying to test.

**Comment by mallenexpensify** [comment]
@cead22 when you have a min can you let me know if the screenshots from testing above _make sense_?

**Comment by mallenexpensify** [comment]
@cead22 , when ya have a min, I think/hope the [above details ](https://github.com/Expensify/App/issues/23679#issuecomment-1713294455)are what you were asking for.  Thx

**Comment by cead22** [comment]
That looks like how it's supposed to be working, given there's an AddEmojiReaction request in the queued, followed by a RemoveEmojiReaction. I'm gonna test locally no

**Comment by cead22** [comment]
I couldn't reproduce with the steps in the issue description. After reconnecting I saw 2 AddEmojiReaction requests that failed, then a third AddEmojiReaction request that succeeded, and then a RemoveEmojiReaction that succeeded. This was the case for both accounts in the two different browser windows

**Comment by Nathan-Mulugeta** [comment]
Hey team I was retesting this now and I am able to reproduce this consistently. I will attach a screen recording down below. cc @cead22 @mallenexpensify 


https://github.com/Expensify/App/assets/111440031/953071a5-7a97-4052-962f-4316a84ae824

**Comment by Nathan-Mulugeta** [comment]
Hey @aimane-chnaif I just retested this again and YES I am still able to reproduce this.

**Comment by mallenexpensify** [comment]
Yeah, I keep meaning to re-test on my end too, I was having trouble logging in with my test account last week.

**Comment by mallenexpensify** [comment]
https://github.com/Expensify/App/assets/22508304/c8e7d9a0-319d-46fd-a185-b315751ac738

**Comment by mallenexpensify** [comment]
Gonna post in #expensify-open-source to get some eyes

**Comment by DylanDylann** [comment]
<img width="1440" alt="Screenshot 2023-11-04 at 10 50 49" src="https://github.com/Expensify/App/assets/141406735/adcc75fd-ca41-4f34-a619-bbcdbd45042e">

The pusher pushes removed reaction data before pushing add reaction data

**Comment by barros001** [comment]
I've spent some time trying to figure this one out and here's what I'm at so far.

I was able to reproduce it on staging with two separate Chrome windows. To be this looks like a race condition, that's why it works sometimes, and sometimes it doesn't. Once I was able to reproduce I looked through the logs on the affected Chrome window and started tracing the requests and responses. Here's the sequence:

* https://staging.expensify.com/api?command=AddEmojiReaction: Adds emoji for user 15921625
* https://staging.expensify.com/api?command=RemoveEmojiReaction: Removes emoji for user 15921625
* https://staging.expensify.com/api?command=ReconnectApp: This response contains several merge actions, which are for some reason duplicate:
  * Adds emoji for user 15921625
  * Adds emoji for user 15912050 (from the other window)
  * Removes emoji for user 15921625
  * Removes emoji for user 15912050
  * Adds emoji for user 15912050
  * Adds emoji for user 15921625
  * Removes emoji for user 15912050
  * Remove emoji for user 15921625
* https://staging.expensify.com/api?command=OpenReport: re-opens the report, but this api does not have any reactions data, so shouldn't affect anything

In the meantime, pusher is active as well:

* OnyxUpdateManager] Applying update type: pusher with lastUpdateID: 27743511: this one adds emoji for user 15912050 (the other window)
* [OnyxUpdateManager] Applying update type: pusher with lastUpdateID: 27743515: and this finally removes emoji for user 15912050

The emoji that's left on the screen belongs to user 15912050, the other screen. Tracking all the merge operations, as long as the `reconnectApp` batch actions are applied in sequence (which they should otherwise everything in the app would break), then I do not see any possible sequence of events that would leave this in this state. The last operations are always removing the emoji. The API requests are in sequence, so it will end with a remove emoji action. Reconnect app is also in sequence, it would end with a remove emoji. Pusher the same. Even if these messages end up mixed together, the last one is always removing an emoji.

One thing that occurred to be is that maybe the app is ignoring the last pusher, maybe for being out of sequence or something. Looking for this I found this code:

```js
function apply({lastUpdateID, type, request, response, updates}: OnyxUpdatesFromServer): Promise<void | Response> | undefined {
    console.debug(`[OnyxUpdateManager] Applying update type: ${type} with lastUpdateID: ${lastUpdateID}`, {request, response, updates});

    if (lastUpdateID && lastUpdateIDAppliedToClient && Number(lastUpdateID) < lastUpdateIDAppliedToClient) {
        console.debug('[OnyxUpdateManager] Update received was older than current state, returning without applying the updates');
        return Promise.resolve();
    }
    if (lastUpdateID && (lastUpdateIDAppliedToClient === null || Number(lastUpdateID) > lastUpdateIDAppliedToClient)) {
        Onyx.merge(ONYXKEYS.ONYX_UPDATES_LAST_UPDATE_ID_APPLIED_TO_CLIENT, Number(lastUpdateID));
    }
    if (type === CONST.ONYX_UPDATE_TYPES.HTTPS && request && response) {
        return applyHTTPSOnyxUpdates(request, response);
    }
    if (type === CONST.ONYX_UPDATE_TYPES.PUSHER && updates) {
        return applyPusherOnyxUpdates(updates);
    }
}
```

Which tells me I would've seen a message saying that pusher message was ignored, which I did not. Looking closer at this function, there's one tiny detail that caught my attention.

If `lastUpdateID` === `lastUpdateIDAppliedToClient`, the message is ignored without any messages to console. That's because the first if checks if `lastUpdatedID > lastUpdateIDAppliedToClient` and the second if `lastUpdatedAt < lastUpdateIDAppliedToClient`.

It does make some sense because if they are the same that means we already applied that update, but just wanted to mention as there's the only thing that I could see that could remotely explain what's happening here. Updates are being sent around, maybe out of order, and we end up ignoring that very last push message because we thought we already applied it, when maybe we have not 🤷

**Comment by barros001** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

On a chat between two accounts, if both are offline, and both add an emoji to a message, and then remove, once they go offline, one of the accounts will still display the emoji attached to the message where it shouldn't.

### What is the root cause of that problem?

The root cause is how pusher events are handled. After looking very closely at the logs and adding some logs myself, I found out that pusher events, although they are received and enqueued in order, they are not guaranteed to be processed in order. Here's what's happening on this example. See the images below:

![Screenshot 2023-11-04 at 4 00 22 PM](https://github.com/Expensify/App/assets/765936/fa988279-1326-4b6a-84ee-be14981add00)
![Screenshot 2023-11-04 at 4 00 27 PM](https://github.com/Expensify/App/assets/765936/45816376-42e1-46e4-b7fa-353842ea0964)
![Screenshot 2023-11-04 at 4 00 32 PM](https://github.com/Expensify/App/assets/765936/1f03d9d2-587e-4195-9853-509240f5894b)
![Screenshot 2023-11-04 at 4 00 37 PM](https://github.com/Expensify/App/assets/765936/759bd337-0ffc-4235-b73f-4c0636beddd3)
![Screenshot 2023-11-04 at 4 00 49 PM](https://github.com/Expensify/App/assets/765936/5c2e159a-786c-43ba-b2e6-358b1cc9fe6f)
![Screenshot 2023-11-04 at 4 01 00 PM](https://github.com/Expensify/App/assets/765936/9301cb27-41fb-4c6f-9542-e6f9c1030fec)
![Screenshot 2023-11-04 at 4 01 07 PM](https://github.com/Expensify/App/assets/765936/98de5b35-1a71-4b0d-9f66-73bb82d92fd6)

This is the real time sequence of events that took place once I went back online (my line numbers might be slightly off because I added some debug).

* Got an event from the pusher, to add a +1 to the message (src/libs/actions/OnyxUpdates.ts:62 - apply)
* Enqueued the +1 to the message in a promise (src/libs/actions/OnyxUpdates.ts:47 - applyPusherOnyxUpdates)
* Got another event from pusher, this time to remove +1 from the message (src/libs/actions/OnyxUpdates.ts:62 - apply). Note that this happened before the first event was processed, which is OK (it should at least)
* Enqueued the "remove +1 from the message" in a promise (src/libs/actions/OnyxUpdates.ts:47 - applyPusherOnyxUpdates)

Now that's where it gets interesting. This is what happened:

* Processed "remove +1 from message" event first (src/libs/actions/User.js:531)
* Processed "add +1 to message" later (src/libs/actions/User.js:531)

You can see on the images highlighted in blue as well as in the "Done applying" messages right after then. They are out of order! So how did this happen? The messages have been received in sequence, enqueued in promises in the right sequence, but the promised resolved out of order. Here's why:

```js
            return SequentialQueue.getCurrentRequest().then(() => {
                // If we don't have the currentUserAccountID (user is logged out) we don't want to update Onyx with data from Pusher
                if (!currentUserAccountID) {
                    return;
                }


                const onyxUpdatePromise = Onyx.update(pushJSON);
                triggerNotifications(pushJSON);

                // Return a promise when Onyx is done updating so that the OnyxUpdatesManager can properly apply all
                // the onyx updates in order
                return onyxUpdatePromise;
            });
```

This is the event subscriber (src/libs/actions/User.js:531). Before processing the event, it will call `getCurrentRequest`, which in itself returns a promise. If for whatever reason this takes more than one cycle to resolve, which with all the stuff going on in the background, it's very much likely to, events will get out of order! Imagine it's processing "add +1", call `getCurrentRequest`, it does not return right away, JS will move on and get back to it later. When it moves on, it will eventually get to "remove +1", which will also call `getCurrentRequest`, except that now it will return in the same cycle. Event got processed out of order!

The precise reasons as to why this happens hard to pinpoint. There's a lot going on on the "current request". When you come online there's apis being fired, pusher events being received, all at once. Either way, the entire process of processing pusher events relies on several level deep of promises and all it takes is one of them needing an extra cycle to resolve (`getCurrentRequest`).

Bottom line, pusher events processing code assumes that once you enqueue the promises, they will be executed in sequence, and this turns out to be a flawed assumption.

### What changes do you think we should make in order to solve the problem?

I propose that we handle this on the `TYPE.MULTIPLE_EVENT_TYPE.ONYX_API_UPDATE` subscriber within src/libs/actions/User.js:527. Instead of assuming `getCurrentRequest` will resolve in the right order, we remove this assumption. For this we introduce a `stack` of events. Whenever a new `pushJSON` is received, we push it into a stack, before calling `getCurrentRequest`. Once this promise resolves, we `shift` the stack and get the oldest entry from it. That way, even if it ends up resolving out of order, we still get the events in the right order. Here's the suggested code:

```js
function subscribeToUserEvents() {
    const pusherEvents = [];
...
...
...

    PusherUtils.subscribeToMultiEvent(Pusher.TYPE.MULTIPLE_EVENT_TYPE.ONYX_API_UPDATE, (inputPushJSON) => {
            pusherEvents.push(inputPushJSON);

            return SequentialQueue.getCurrentRequest().then(() => {
                const pushJSON = pusherEvents.shift();

                // If we don't have the currentUserAccountID (user is logged out) we don't want to update Onyx with data from Pusher
                if (!currentUserAccountID) {
                    return;
                }

                const onyxUpdatePromise = Onyx.update(pushJSON);
                triggerNotifications(pushJSON);

                // Return a promise when Onyx is done updating so that the OnyxUpdatesManager can properly apply all
                // the onyx updates in order
                return onyxUpdatePromise;
            });
        }
    );

...
...
...
}
```

I tested it locally and was able to see a log where the inner code (where we shift pusherEvents) got a different event than what it has in its `inputPushJSON`, meaning it was processed out of sequence but still managed to process the right events in the right order. The UI issue disappeared after that.




### What alternative solutions did you explore? (Optional)

**Reminder:** Please use plain English, be brief and avoid jargon. Feel free to use images, charts or pseudo-code if necessary. Do not post large multi-line diffs or write walls of text. Do not create PRs unless you have been hired for this job.

**Comment by barros001** [comment]
@aimane-chnaif from what I could gather the root cause is the fact that we have several levels of nested promises when a pusher event is processed.

* `applyPusherOnyxUpdates` creates one promise for each pusher event
* each of these promises will ultimately call `SequentialQueue.getCurrentRequest()` which is another promise
* currentRequest promise waits on `Request.processWithMiddleware` promise
* processWithMiddleware could potentially have more promises

With a lot going on in the background when you come back online (queued api requests, pusher messages, etc) I don't see how we can guarantee the order of pusher events because all these nested promises will resolve in unpredictable sequence.

I don't fully understand why `SequentialQueue.getCurrentRequest().then(() => {...})` is needed within `subscribeToMultiEvent` but this is the main culprit here. But I have to assume there's a good reason for this call to be there and in that case I can't see a way to guarantee the sequence of pusher events without pushing them into a FIFO queue and processed them in sequence.

I see your point that it looks like a workaround and I do somewhat agree with it, but I'm yet to think of another solution for this problem (I'll give it some more thoughts). If anyone else has more ideas now that the real root cause was discovered, I'd be happy to hear.

**Comment by barros001** [comment]
## Proposal

[Updated](https://github.com/Expensify/App/issues/23679#issuecomment-1793548918)

**Comment by barros001** [comment]
@aimane-chnaif I spent some more time looking into this and I believe I found the real root cause and a proper way to solve this. let me know your thoughts on the updated proposal.

**Comment by barros001** [comment]
## Proposal

[Updated](https://github.com/Expensify/App/issues/23679#issuecomment-1793548918)

It just occurred to me that the second part of my proposal was not necessary, I was wrapping a promise into another promise which was redundant. The final proposal is much simpler and changes smaller.

**Comment by tgolen** [comment]
I love that root-cause-analysis! Thank you for being so thorough. I understand the problem and it's extremely similar to a recent change that was made in Onyx to ensure that multiple database writes were handled in the correct order. You can see the PR here: https://github.com/Expensify/react-native-onyx/pull/384

I am having a hard time wrapping my head around how the solution is supposed to fix it exactly. When I have an array of things that each need a promise, and I want to run them in sequence (one after the other), I usually turn to something that looks more like this: https://stackoverflow.com/questions/43082934/how-to-execute-promises-sequentially-passing-the-parameters-from-an-array

**Comment by tgolen** [comment]
but in general, the proposal makes sense 👍 Make the promises happen sequentially.

**Comment by barros001** [comment]
> I am having a hard time wrapping my head around how the solution is supposed to fix it exactly.

@tgolen the thing with this issue is that we have two problems here:

* `applyPusherOnyxUpdates` could be called with more than one `OnyxUpdateEvent` and we want these to be executed in sequence.
* at the same time, `applyPusherOnyxUpdates` could be called **multiple times**, in parallel.

If `applyPusherOnyxUpdates` is called again while promises from a previous call are still executing, then we need to add them to the promise chain so that we wait until them all are finished before executing the new batch. Otherwise we'll just start a new promise chain in parallel, not solving this problem.

That's why my proposal has a promise at the top of the file (naming can be improved I admit), we can't deal with this locally inside of `applyPusherOnyxUpdates` due to the possible multiple parallel calls to it.

```js
let pusherEventsQueuePromise = Promise.resolve();
```

All new promises processed by `applyPusherOnyxUpdates` will have to be chained in this one as to avoid processing them in parallel.

```js
    for (const update of updates) {
        pusherEventsQueuePromise = pusherEventsQueuePromise.then(() => PusherUtils.triggerMultiEventHandler(update.eventType, update.data));
    }
```

The code above is responsible for creating this chain of promises handling both scenarios described earlier (multiple events in one call and multiple calls). 

```
promise
  .then(()=>{...})
  .then(()=>{...})
```
is the same as:

```
promise = promise.then(()=>{...}); 
promise = promise.then(()=>{...});
```

Let me know if this makes sense.

**Comment by tgolen** [comment]
Sure, I'll assign it to myself and to @barros001 for the proposal

**Comment by barros001** [comment]
Thanks for accepting my proposal, I'll have a PR up for review tomorrow.

**Comment by barros001** [comment]
[PR](https://github.com/Expensify/App/pull/31144) is up for review.

@tgolen I ended up refactoring the `for...` block to use `reduce` instead based on the link you provided.

**Comment by mallenexpensify** [comment]
@aimane-chnaif can you please complete the checklist above?  Thx

**Comment by mallenexpensify** [comment]
Bug reporter: @Nathan-Mulugeta paid $250 via Upwork (report was from wayyyyyy back)
Contributor: @barros001 paid $1000 via Upwork
Contributor+: @aimane-chnaif paid $1000 via Upwork


Thanks @aimane-chnaif I agree with
> I think this bug existed for long time. Not able to find offending PR.

Created the testrail update GH
https://github.com/Expensify/Expensify/issues/340437

**Comment by nyomanjyotisa** [comment]
The shared Pusher promise chain introduced in #31144 lacked rejection recovery, causing #100174: one failed apply prevented subsequent Pusher updates from running for the rest of the session. Fixed in #100594 with chain recovery and watermark continuity protection.
